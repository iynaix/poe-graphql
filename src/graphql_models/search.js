import isEmpty from "lodash/isEmpty"
import mapKeys from "lodash/mapKeys"

import { mergeMongoQueries } from "./utils"

export const searchNumeric = (fieldName, fieldValue) => {
    return { [fieldName]: mapKeys(fieldValue, (_, k) => k.replace("_", "$")) }
}

// searchNumeric is just a superset of searchBoolean!
export const searchBoolean = searchNumeric

// utility function for handling search operations
const _stringOperator = ([op, val]) => {
    switch (op) {
        case "_eq":
            return { $eq: val }
        case "_ne":
            return { $ne: val }
        case "_ieq":
            return { $regex: val, $options: "i" }
        case "_regex":
            return { $regex: val }
        case "_iregex":
            return { $regex: val, $options: "i" }
        case "_contains":
            return { $regex: val }
        case "_icontains":
            return { $regex: val, $options: "i" }
        case "_startswith":
            return { $regex: `^${val}` }
        case "_istartswith":
            return { $regex: `^${val}`, $options: "i" }
        case "_endswith":
            return { $regex: `${val}$` }
        case "_iendswith":
            return { $regex: `${val}$`, $options: "i" }
        case "_in":
            return { $in: val }
        case "_nin":
            return { $nin: val }
        default:
            throw Error(`Invalid string operation: ${op}`)
    }
}

export const searchString = (fieldName, fieldValue) => {
    const mongoParams = Object.entries(fieldValue).map(_stringOperator)

    return {
        $and: mongoParams.map((v) => ({
            [fieldName]: v,
        })),
    }
}

export const searchWhereRecursive = (searchParams, searchFunc) => {
    if (!searchParams) {
        return {}
    }

    const queries = [searchFunc(searchParams)]

    // handle special mongodb operators
    const logicalOps = { _or: "$or", _and: "$and", _not: "$nor" }
    Object.entries(logicalOps).forEach(([op, mongoOp]) => {
        if (op in searchParams) {
            queries.push({
                [mongoOp]: searchParams[op].map((v) => searchWhereRecursive(v, searchFunc)),
            })
        }
    })

    return mergeMongoQueries(...queries)
}
