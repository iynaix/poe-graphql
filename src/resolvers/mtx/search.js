export const searchWatchlist = (search, watchlistIds = []) => {
    if (search.watchlist !== undefined && watchlistIds.length) {
        return { id: { [search.watchlist ? "$in" : "$nin"]: watchlistIds } }
    }
}

export const searchOwned = (search, ownedNames = []) => {
    if (search.owned !== undefined && ownedNames.length) {
        return { name: { [search.owned ? "$in" : "$nin"]: ownedNames } }
    }
}
