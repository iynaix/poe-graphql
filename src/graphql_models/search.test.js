import { searchNumeric, searchString, searchWhereRecursive } from "./search"
import { createMongoFilter } from "."

describe("searchNumeric", () => {
    test("single condition", () => {
        expect(searchNumeric("a", { _eq: 1 })).toMatchSnapshot()
    })

    test("multiple conditions", () => {
        expect(searchNumeric("a", { _eq: 1, _lt: 2 })).toMatchSnapshot()
    })
})

describe("searchString", () => {
    test("single condition", () => {
        expect(searchString("a", { _eq: "1" })).toMatchSnapshot()
    })

    test("multiple conditions", () => {
        expect(searchString("a", { _eq: "1", _iregex: "2" })).toMatchSnapshot()
    })
})

const multiFields = {
    a: { type: "String" },
    b: { type: "String" },
}

describe("complex queries", () => {
    test("empty field searches should do nothing", () => {
        const fields = { a: { type: "String" } }
        const where = { a: {} }

        expect(createMongoFilter(fields)(where)).toStrictEqual({})
    })

    test("multiple string searches should not override $and", () => {
        const where = { a: { _eq: "1" }, b: { _eq: "2" } }

        expect(createMongoFilter(multiFields)(where)).toMatchSnapshot()
    })

    test("non-recursive _or search", () => {
        const where = {
            _or: [{ a: { _eq: "1" } }, { b: { _eq: "2" } }],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("nested _or search", () => {
        const where = {
            _or: [{ a: { _eq: "1" }, _or: [{ b: { _eq: "2" } }] }],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("recursive _or search", () => {
        const where = {
            _or: [
                { a: { _eq: "1" } },
                { b: { _eq: "2" } },
                { _or: [{ a: { _eq: "1" } }, { b: { _eq: "2" } }] },
            ],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("non-recursive _and search", () => {
        const where = {
            _and: [{ a: { _eq: "1" } }, { b: { _eq: "2" } }],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("nested _and search", () => {
        const where = {
            _and: [{ a: { _eq: "1" }, _and: [{ b: { _eq: "2" } }] }],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("recursive _and search", () => {
        const where = {
            _and: [
                { a: { _eq: "1" } },
                { b: { _eq: "2" } },
                { _and: [{ a: { _eq: "1" } }, { b: { _eq: "2" } }] },
            ],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("non-recursive _not search", () => {
        const where = {
            _not: [{ a: { _eq: "1" } }, { b: { _eq: "2" } }],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("nested _not search", () => {
        const where = {
            _not: [{ a: { _eq: "1" }, _not: [{ b: { _eq: "2" } }] }],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })

    test("recursive _not search", () => {
        const where = {
            _not: [
                { a: { _eq: "1" } },
                { b: { _eq: "2" } },
                { _not: [{ a: { _eq: "1" } }, { b: { _eq: "2" } }] },
            ],
        }
        const searchFunc = createMongoFilter(multiFields)

        expect(searchWhereRecursive(where, searchFunc)).toMatchSnapshot()
    })
})
