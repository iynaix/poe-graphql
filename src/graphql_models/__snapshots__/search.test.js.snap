// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex queries multiple string searches should not override $and 1`] = `
Object {
  "$and": Array [
    Object {
      "a": Object {
        "$eq": "1",
      },
    },
    Object {
      "b": Object {
        "$eq": "2",
      },
    },
  ],
}
`;

exports[`complex queries nested _and search 1`] = `
Object {
  "$and": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
        Object {
          "$and": Array [
            Object {
              "b": Object {
                "$eq": "2",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`complex queries nested _not search 1`] = `
Object {
  "$nor": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
      "$nor": Array [
        Object {
          "$and": Array [
            Object {
              "b": Object {
                "$eq": "2",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`complex queries nested _or search 1`] = `
Object {
  "$or": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
      "$or": Array [
        Object {
          "$and": Array [
            Object {
              "b": Object {
                "$eq": "2",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`complex queries non-recursive _and search 1`] = `
Object {
  "$and": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "b": Object {
            "$eq": "2",
          },
        },
      ],
    },
  ],
}
`;

exports[`complex queries non-recursive _not search 1`] = `
Object {
  "$nor": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "b": Object {
            "$eq": "2",
          },
        },
      ],
    },
  ],
}
`;

exports[`complex queries non-recursive _or search 1`] = `
Object {
  "$or": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "b": Object {
            "$eq": "2",
          },
        },
      ],
    },
  ],
}
`;

exports[`complex queries recursive _and search 1`] = `
Object {
  "$and": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "b": Object {
            "$eq": "2",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "$and": Array [
            Object {
              "a": Object {
                "$eq": "1",
              },
            },
          ],
        },
        Object {
          "$and": Array [
            Object {
              "b": Object {
                "$eq": "2",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`complex queries recursive _not search 1`] = `
Object {
  "$nor": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "b": Object {
            "$eq": "2",
          },
        },
      ],
    },
    Object {
      "$nor": Array [
        Object {
          "$and": Array [
            Object {
              "a": Object {
                "$eq": "1",
              },
            },
          ],
        },
        Object {
          "$and": Array [
            Object {
              "b": Object {
                "$eq": "2",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`complex queries recursive _or search 1`] = `
Object {
  "$or": Array [
    Object {
      "$and": Array [
        Object {
          "a": Object {
            "$eq": "1",
          },
        },
      ],
    },
    Object {
      "$and": Array [
        Object {
          "b": Object {
            "$eq": "2",
          },
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "$and": Array [
            Object {
              "a": Object {
                "$eq": "1",
              },
            },
          ],
        },
        Object {
          "$and": Array [
            Object {
              "b": Object {
                "$eq": "2",
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`searchNumeric multiple conditions 1`] = `
Object {
  "a": Object {
    "$eq": 1,
    "$lt": 2,
  },
}
`;

exports[`searchNumeric single condition 1`] = `
Object {
  "a": Object {
    "$eq": 1,
  },
}
`;

exports[`searchString multiple conditions 1`] = `
Object {
  "$and": Array [
    Object {
      "a": Object {
        "$eq": "1",
      },
    },
    Object {
      "a": Object {
        "$options": "i",
        "$regex": "2",
      },
    },
  ],
}
`;

exports[`searchString single condition 1`] = `
Object {
  "$and": Array [
    Object {
      "a": Object {
        "$eq": "1",
      },
    },
  ],
}
`;
